appsrc/dcfile/dccsdump.man
.TH DCCSDUMP 1 "28 February 2026" "DICOM PS3" "DICOM PS3 - Describe DICOM coded sequence content"
.SH NAME
dccsdump \- ACR/NEMA DICOM PS3 ... DICOM PS3 - Describe DICOM coded sequence content
.SH SYNOPSIS
.HP 10
.B dccsdump
.so man1/gen.so
[
.B \-v|verbose
]
[
.B \-vv|veryverbose
]
[
.B \-filename
]
[
.B \-identifier
]
.so man1/optin.so
.SH DESCRIPTION
.LP
.B dccsdump
reads the named dicom input file and describes any Code Sequences (defined by the presence of Coding Scheme Designator) contained therein.
.SH OPTIONS
The attribute values, description and verbose output go to standard error.
.PP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-verbose
.RS
Describe the parsing process; does not actually do anything at present.
.RE
.TP
.B \-veryverbose
.RS
Display byte offset from file start and message start, in hexadecimal, and dump contents of the DICOM dataset while reading.
.RE
.TP
.B \-filename
.RS
Show the name of the file supplied in the arguments; a hyphen will be reported if no filename was supplied.
.RE
.TP
.B \-identifier
.RS
Preceed each sequence with the dot delimited numeric identifier of the sequence attribute containing the coded content relative to the root of the nested sequence item tree, and the item number in the sequence in brackets.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% dccidump -identifier wsi.dcm
.RE
		1.81[1].6[1].1: Specimen Preparation Step Content Item Sequence
.RE
1.87: Container Type Code Sequence = (433466003,SCT,"Microscope slide")
.RE
	1.89[1].1: Primary Anatomic Structure Sequence = (10003008,SCT,"Unspecified")
.RE
			1.89[1].6[1].1[1].2: Concept Name Code Sequence = (121041,DCM,"Specimen Identifier")
.RE
			1.89[1].6[1].1[2].2: Concept Name Code Sequence = (111701,DCM,"Processing type")
.RE
			1.89[1].6[1].1[2].3: Concept Code Sequence = (127790008,SCT,"Staining")
.RE
			1.89[1].6[1].1[3].2: Concept Name Code Sequence = (424361007,SCT,"Using substance")
.RE
			1.89[1].6[1].1[3].3: Concept Code Sequence = (12710003,SCT,"hematoxylin stain")
.RE
			1.89[1].6[1].1[4].2: Concept Name Code Sequence = (424361007,SCT,"Using substance")
.RE
			1.89[1].6[1].1[4].3: Concept Code Sequence = (36879007,SCT,"water soluble eosin stain")
.RE
	1.101[1].1: Illumination Type Code Sequence = (111744,DCM,"Brightfield illumination")
.RE
	1.101[1].5: Illumination Color Code Sequence = (414298005,SCT,"Full Spectrum")
.RE
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dcintro(1) ,
.BR dcdump(1) ,
.BR dccidump(1)
.SH AUTHOR
Copyright (c) 1993-2026, David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
260701:	libsrc/standard/strval/base.tpl: Add 99PRAMANA CSD encountered

260627:	libsrc/standard/elmdict/other.tpl: Add more Pramana WSI private data elements encountered

260624:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check lengths of Extended Offset Table matches NumberOfFrames (000641)

260624:	appsrc/dcfile/dcinfo.cc: Correct detection of empty Extended Offset Table (000640)

260624:	appsrc/dcfile/dcinfo.cc: Report whether instance of concatenation or not (000639)

260616:	libsrc/standard/condn.tpl: Do not complain about presence of Referenced Series Sequence in Common Instance Reference Module when images referenced from Conversion Source Attributes Sequence (000638)

260613:	libsrc/standard/condn.tpl: Check whether or not IOD permits Segmented Lookup Table Data for other than images (000637)

260613:	libsrc/standard/condn.tpl: Do not complain about presence of Referenced Series Sequence in Common Instance Reference Module when images referenced from Advanced Blending Sequence (000636)

260603:	libsrc/standard/elmdict/dicom3.tpl: Add Display URI data element (CP 2595)

260324:	libsrc/standard/elmdict/siemens.tpl: Remove duplicate MR private data elements and add some from XA60A conformance statement and encountered images

260324:	libsrc/standard/condn.tpl,module/vl.tpl: Add Acquisition and Pyramid UID requirements to WG26SP2025 profile (CP-LAB-279)

260324:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Report invalid presence of PyramidUID for LABEL and OVERVIEW image types (000577)

260321:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for metal artifact reduction (CP 2575)

260228:	libsrc/standard/module/base.tpl,vl.tpl: Detect missing PrimaryAnatomicStructureSequence in WG26SP2025 profile (000635)

260228:	Imakefile,dcfile/Imakefile,appsrc/dcfile/dccsdump.cc,dccsdump.man,dccidump.man,dcdump.man: Add utility to dump code sequence items (000634)

260228:	appsrc/dcfile/dccidump.cc,dcsrdump.cc: Show Long Code Value or URN Code Value used when Code Value not present (000633)

260228:	appsrc/dcfile/dccidump.cc: Show numeric values in non-SR content items (NUMERIC rather than NUM) (000632)

260201:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for new defined terms for Dose Type to RT Dose Module (CP 2535)

260123:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Add mechanism for handling sensitive images (CP 1982)

260117:	libsrc/standard/condn.tpl,sopcl.tpl,iodcomp,module,strval/base.tpl: Add DICOM Encapsulation of OBJ Models for 3D Manufacturing and Virtual Reality (Sup 208)

251221:	libsrc/standard/condn.tpl: Check for SCT human species as well as SRT (000630)

251020: appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Warn when DimensionIndexPointer Attribute not present in any Shared or Per-Frame Functional Group (000629)

251020: libsrc/standard/condn.tpl,module/base.tpl: Check DimensionOrganizationType is absent as alternative to not TILED_FULL for DimensionIndexSequence condition (CP 1822)

250914:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Add Scan Spot Gantry and Patient Support Angle to Ion Beam Record (CP 2504)

250907:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Add Dose Reporting Mode and Dose Calculation Model to RT Dose (CP 2428)

250817: libsrc/standard/condn.tpl: Do not report missing PlanePositionSequence and PlaneOrientationSequence for WSI Parametric Map (000628)

250817: libsrc/standard/strval/base.tpl,ct.tpl,mr.tpl: Update Pixel Contrast Image/Frame Type Value 4 to use QUANTITY, expecting details to be in RWVM Quantity Definition Sequence (CP 1700)

250723:	appsrc/dcfile/dcinfo.cc,.man,Imakefile: Add utility to describe characteristics of images esp. re. multi-frame, enhanced, encapsulated and presence or absence of offset tables (000627)

250715:	appsrc/dcfile/dciodvfy.cc: Check consistency of ImagedVolumeWidth,Height with TotalPixelMatrixColumns,Rows and PixelSpacing (000626)

250713:	libsrc/standard/elmdict/other.tpl: Add PixelMed original file message digest private data elements

250709:	libsrc/standard/module/rt.tpl: Relax Type 1 RT Structure Set requirements (CP 2445)

250709:	libsrc/standard/elmdict/base.tpl: Add data element for Add RT specific plan approvals (CP 2427)

250605:	libsrc/standard/transyn.tpl,strval/base.tpl: Add HTJ2K Transfer Syntaxes and string values (Sup 235)

250605:	libsrc/standard/transyn.tpl,strval/base.tpl: Add JPEG-XL Transfer Syntaxes and string values (Sup 232)

250605:	libsrc/include/dctool/vr.h,src/dctool/attrmxrd.cc: Check VR is OB for encapsulated compressed PixelData (000624)

250525:	appsrc/dcfile/dciodvfy.cc,dciodvfy.man,libsrc/src/dctool/attrmxls.cc,libsrc/support/module.awk: Check only first item of Referenced Image Navigation Sequence by default (000623)

250524:	appsrc/dcfile/dciodvfy.cc,dciodvfy.man,libsrc/include/dctool/attrmxls.h,iodcomp.h,module.h,libsrc/src/dctool/attrmxls.cc,libsrc/support/iodcomp.awk,module.awk: Check only first item of Per-Frame Functional Groups Sequence by default, without specific command line option to do them all (000621)

250523:	libsrc/standard/strval/base.tpl: Add 99Visiopharm coding scheme

250523:	libsrc/standard/iodcomp,module,strval/vl.tpl: Add dciodvfy WSI and ANN profile WG26SP2025 to check for WG 26 2025 Spring Connectathon requirements

250523:	libsrc/standard/elmdict/other.tpl: Add observed Google converted WSI private data elements

250523:	libsrc/standard/elmdict/other.tpl: Add observed Agfa private data element not in conformance statement

250523:	libsrc/standard/elmdict/other.tpl: Add observed 3DHISTECH private data elements not in conformance statement

250523:	libsrc/standard/elmdict/other.tpl: Add Leica Aperio WSI private data elements from Conformance Statement for v1.4

250522:	libsrc/src/locale/mesgtext.tpl,appsrc/dcfile/dciodvfy.cc: Check for instance UIDs that use NEMA root but are not standard well-known instances (000620)


appsrc/dcfile/dccsdump.cc
static const char *CopyrightIdentifier(void) { return "@(#)dccidump.cc Copyright (c) 1993-2024, David A. Clunie DBA PixelMed Publishing. All rights reserved."; }
#include "attrmxls.h"
#include "mesgtext.h"
#include "dcopt.h"
#include "elmconst.h"
#include "attrval.h"

static inline void
indent(TextOutputStream &log,unsigned depth,bool showtabs,bool showgt) {
	unsigned d=depth;
	if (showtabs) while (d--) log << "\t";
	d=depth;
	if (showgt) while (d--) log << ">";
}

static void
writeTagNameToLog(Tag tag,ElementDictionary *dict,TextOutputStream &log) {
	const char *desc = NULL;
	if (dict) {
		desc = dict->getDescription(tag);
	}
	if (desc && strlen(desc) > 0 && strcmp(desc,"?") != 0) {
		log << desc;
	}
}

static void
writeTagNameToLog(Attribute *a,ElementDictionary *dict,TextOutputStream &log) {
	writeTagNameToLog(a->getTag(),dict,log);
}

static void
logStringValueOfAttribute(Attribute *a,TextOutputStream &log) {
	if (a) {
		char *v = AttributeValue(a);
		if (v) {
			log << v;
			delete[] v;
		}
	}
}

static void
processCodeSequenceItem(AttributeList *list,TextOutputStream &log) {
	log << "(";
	{
		Attribute *a = (*list)[TagFromName(CodeValue)];
		if (!a) {	// (000633)
			a = (*list)[TagFromName(LongCodeValue)];
			if (!a) {
				a = (*list)[TagFromName(URNCodeValue)];
			}
		}
		logStringValueOfAttribute(a,log);
	}
	log << ",";
	logStringValueOfAttribute((*list)[TagFromName(CodingSchemeDesignator)],log);
	log << "," << "\"" ;
	logStringValueOfAttribute((*list)[TagFromName(CodeMeaning)],log);
	log << "\")";
}

static bool
processCodeSequences(AttributeList *list,TextOutputStream &log,bool verbose,bool veryverbose,unsigned depth,const char *identifierstring)   {
	bool success=true;
	
	int attributeCount = 0;
	AttributeListIterator listi(*list);
	while (!listi) {
		Attribute *a=listi();
		Assert(a);
		char *attributeidentifierstring = NULL;
		if (identifierstring) {
			ostrstream attributeidentifierstream;
			attributeidentifierstream << identifierstring << "." << (attributeCount+1) << ends;
			attributeidentifierstring = attributeidentifierstream.str();
		}
		if (strcmp(a->getVR(),"SQ") == 0) {
			AttributeList **al;
			int n;
			if ((n=a->getLists(&al)) > 0) {
				bool thisAttributeIsCodeSequence = false;
				if (n>0 && (*al[0])[TagFromName(CodingSchemeDesignator)] && a->getTag() != TagFromName(CodingSchemeIdentificationSequence)) {
					thisAttributeIsCodeSequence = true;
					indent(log,depth,true,false/*showgt*/);
					if (attributeidentifierstring) {
						log << attributeidentifierstring << ": ";
					}
					writeTagNameToLog(a,list->getDictionary(),log);
					log << " = ";
				}
				
				int i;
				for (i=0; i<n; ++i) {
					char *childidentifierstring = NULL;
					if (attributeidentifierstring) {
						ostrstream childidentifierstream;
						childidentifierstream << attributeidentifierstring << "[" << (i+1) << "]" << ends;	// distinguish items in brackets from attribute path separated by periods
						childidentifierstring = childidentifierstream.str();
					}
					
					if (thisAttributeIsCodeSequence) {
						if (i > 0) log << ", ";
						processCodeSequenceItem(al[i],log);
					}
					
					success&=processCodeSequences(al[i],log,verbose,veryverbose,depth+1,childidentifierstring);
					if (childidentifierstring) {
						delete [] childidentifierstring;
					}
				}
				if (thisAttributeIsCodeSequence) log << endl;
				
				delete [] al;
			}
		}
		if (attributeidentifierstring) {
			delete [] attributeidentifierstring;
		}
		++listi;
		++attributeCount;
	}
	return success;
}

static bool
parseCodeSequences(ManagedAttributeList &list,TextOutputStream &log,bool verbose,bool veryverbose,bool showidentifier)
{
	bool result = processCodeSequences(&list,log,verbose,veryverbose,0,showidentifier ? "1" : NULL);
	return result;
}

int
main(int argc, char *argv[])
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);

	bool veryverbose=options.get("veryverbose") || options.get("vv");
	bool verbose=options.get("verbose") || options.get("v") || veryverbose;
	bool showfilename=options.get("filename");
	bool showidentifier=options.get("identifier");

	dicom_input_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);

	cerr << dicom_input_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();

	if (!dicom_input_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< " [-v|-verbose|-vv|-veryverbose]"
			<< " [-filename]"
			<< " [-identifier]"
			<< " [" << MMsgDC(InputFile) << "]"
			<< " <" << MMsgDC(InputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream log(cerr);
	
	if (showfilename) {
		const char *filenameused = input_opener.getFilename();
		cerr << "Filename: \"" << (filenameused && strlen(filenameused) > 0 ? filenameused : "-") << "\"" << endl;
	}

	if (veryverbose) log << "******** While reading ... ********" << endl; 
	list.read(din,false/*newformat*/,&log,veryverbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

	const char *errors=list.errors();
	if (errors) log << errors << flush;
	if (!list.good()) {
		log << EMsgDC(DatasetReadFailed) << endl;
		success=false;
	}

	//if (veryverbose) log << "******** As read ... ********" << endl; 
	//log << list;
	//list.write(log,veryverbose);

	{
		bool parseSuccess = parseCodeSequences(list,log,verbose,veryverbose,showidentifier);		// parse regardless of read success or failure
		success=success&&parseSuccess;
	}

	return success ? 0 : 1;
}




